### PR TITLE
Feat: search field for trips and activities

### LINKED ISSUE #244: Search field for activities / trips

### PR DESCRIPTION
**Pull request by eleonoraka**
## Summary

This PR introduces a new search functionality in both the Overview and Activities screens

## Changes

- **Screens/UI:** 

Overview Screen:

Added search functionality in the TopAppBar with a toggle button
Search field replaces the "Your Trips" title when active
Real-time filtering of trips based on trip names
Close button to clear search and restore title

Activities Screen:

Integrated search field 
Maintains separate "Drafts" and "Final" sections while filtering

- **Bug Fixes/Improvements:** closes #244 

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [x] This PR resolves #244 (if applicable).
- [x] Tests have been added for new functionality.
- [x] Screenshots or UI changes have been attached (if applicable).
- [x] Documentation has been updated (if applicable).

##  Testing

- **Manual Testing:**
    - [x] Tested on emulator / physical device.
- **Unit Tests**
    - [x] Added units tests for this

##  Related Issues/Tickets

Closes #244 

## Screenshots (if applicable)

image
image

### CODE REVIEW

**Comment by g0g05arui** [comment]
@eleonoraka please add a reviewer

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt
Comment (1andreea26) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:136]: Could you please use the string resources folder here instead of hardcoded strings? :)

**Comment by 1andreea26** [comment]
In the future, it would be great if you could include a video of the functionalities!

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt
Comment (1andreea26) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:129]: I liked the colour of this TopAppBar more how it was in Figma.. do you think it looks better like this?

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt
Comment (1andreea26) [inline app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:150]: Since our search bar filters the screen in real time, I think you could remove this line, as it seems to be doing nothing. Would there be a reason to keep it?

Comment (1andreea26) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:150]: Or maybe, we should think about the behaviour we want to implement.. Do we want when the user presses Enter to only see the filtered trips? I think that could be cool, wdyt?

**Comment by 1andreea26** [comment]
Regarding the UI, do you think it would look better if the search bars were a bit smaller? To me they seem a bit big, especially in the Activities Screen.

**Comment by 1andreea26** [comment]
The search bars are functioning very well! I just have one suggestion, lmk what you think: After searching for a nonexistent query, I think it would be better to display a message to the user, just like the one we have in the Search Bar for Users (I attach a picture below).  
Screenshot 2024-12-02 at 19 38 17

**Review by 1andreea26 — CHANGES_REQUESTED** [review]
Great effort and work! Thanks for implementing this! I just left some comments and suggestions. Moreover, nit, but could you please add/update documentation for the composables/classes which lack it? For example Overview and Activities, I think it would really help us to start adding them for M3 (I leave it up to you otherwise).

File: app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:104]: nice! this makes the code readable, nice work!

File: app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:154]: It would be nice to have this search bar on the By day activity screen as well, for consistency, wdyt? This screen is the one that pops up when you click on a certain day in the schedule view. Same for the filter button. You could make these a composable and move it to ActivityUtils maybe, import it from there to the ByDay screen. This can also be a separate PR if you prefer, since it touches the filter button as well.

File: app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt
Comment (ioanajianu) [inline app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:141]: I agree with Andreea, the search field is a bit too big, it leaves little room for scrolling for activities. Try reducing the padding for the start? This might be enough, since now there's quite some unused space because of the padding

**Review by ioanajianu — CHANGES_REQUESTED** [review]
Nice work Eleonora! I left some comments, let me know what you think

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt
Comment (eleonoraka) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt:129]: I thought it could frame the main screen in a nice manner and highlight it as that. But I see your point and I'll fix it in the UI/UX PR I'm working on this week.

File: app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt
Comment (eleonoraka) [reply app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:141]: Thanks for the suggestion! I agree with you both. As for the previous suggestions, I'll convert it to an issue and solve it in the UI/UX PR I'm working on.

File: app/src/main/java/com/android/voyageur/ui/overview/Overview.kt
Comment (eleonoraka) [reply app/src/main/java/com/android/voyageur/ui/overview/Overview.kt]: I just deleted it, deleting the search icon form the keyboard which now just shows a check on the Enter button. Thank you for the remark! :)

**Comment by eleonoraka** [comment]
> Great effort and work! Thanks for implementing this! I just left some comments and suggestions. Moreover, nit, but could you please add/update documentation for the composables/classes which lack it? For example Overview and Activities, I think it would really help us to start adding them for M3 (I leave it up to you otherwise).

I agree with you on the documentation but I don't see what it could add to the search bar feature. I think it is already explicit that it is a search field. Could you be a bit more specific please? Do you mean for the whole class?

File: app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt
Comment (ioanajianu) [reply app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:146]: Thanks for your input! Since this fix is directly related to the current PR, I think it would be better to address it here than in the UI/UX one. That PR should be for fixes for PRs that have already been merged. 
As for my other comment, I think that it should be done either here or in a new PR, not in the UI/UX one since it's a feature, and not UI related. What do you think?

Comment (eleonoraka) [reply app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:147]: As requested, I changed the padding and size of the search bar in the Activities screen, closing the issue #272. 
For the search bar in the ByDay screen, I think it could be added in the UI/UX PR since it would not be a new feature but an improvement on the UX for the ByDay screen. Let me know what you think :)

Comment (ioanajianu) [reply app/src/main/java/com/android/voyageur/ui/trip/activities/Activities.kt:147]: Thanks for working on fixing the padding, however for me it still looks like not much changed. Is this the intended behavior?
search bar


 I still think that the search bar on the ByDay screen belongs better in this PR than the UI/UX one (since I was thinking that that one should be more about little improvements, not new working components). But, if you strongly believe that should be the case, I leave it up to you.

**Comment by g0g05arui** [comment]
@eleonoraka 
Thank you for the effort, however I think the design presented here is not similar to the one in the search page. Maybe it would be a good idea to keep the look/feel of the app consistent.

>I'll add it in the next PR.

I think it would be beneficial to add it in this PR as then you would need to redo all the tests and refactoring them could take a while.

Also
>As requested, I changed the padding and size of the search bar in the Activities screen, closing the issue https://github.com/swent-t16/Voyageur/issues/272.

Please do not create issues for every comment / sugestion on this PR as this just clutters the board and they are not needed as this PR is not yet merged into main.

**Comment by 1andreea26** [comment]
> > Great effort and work! Thanks for implementing this! I just left some comments and suggestions. Moreover, nit, but could you please add/update documentation for the composables/classes which lack it? For example Overview and Activities, I think it would really help us to start adding them for M3 (I leave it up to you otherwise).
> 
> I agree with you on the documentation but I don't see what it could add to the search bar feature. I think it is already explicit that it is a search field. Could you be a bit more specific please? Do you mean for the whole class?

Yes, I meant in the whole class, but for now no worries, I am adding them for the Overview in my PR already. It's just to keep in mind in the future, could save us some time before the deadline.

**Comment by 1andreea26** [comment]
> > The search bars are functioning very well! I just have one suggestion, lmk what you think: After searching for a nonexistent query, I think it would be better to display a message to the user, just like the one we have in the Search Bar for Users (I attach a picture below). Screenshot 2024-12-02 at 19 38 17
> 
> Thank you for the suggestion! I agree with you, it would make the user experience better. I'll add it in the next PR.

I would suggest either adding it in this PR, or creating an issue for it :) As you want, but i don't think it's a good idea to mix PRs.

**Review by ioanajianu — COMMENTED** [review]
Thanks for implementing our suggestions! I have a few comments to make:
- nice work in reusing the `NoResultsFound` composable!
- the padding looks better now, thanks for fixing that
- you didn't reply to my comment with the screenshot, it's not clear to me what you decided to do with the search bar for the ByDay screen
- I see that in the UI/UX fixes you made the search bar from the Search screen a separate composable, do you plan to reuse it for trips and activities? To me it would seem nicer if the search bar looks like the one on the Search screen for consistency, but depends what you and @1andreea26 think as well
- please solve the merge conflicts
- the last commit didn't go through the CI, this happens sometimes, but it will be fixed once you push the merge changes. It's important that we have a look at the test coverage as well.

Overall, the search bar works well, thanks for implementing this feature and taking into consideration our suggestions! I'll let Andreea give you the approval once she has a look too.

**Comment by eleonoraka** [comment]
> Thanks for implementing our suggestions! I have a few comments to make:
> 
> * nice work in reusing the `NoResultsFound` composable!
> * the padding looks better now, thanks for fixing that
> * you didn't reply to my comment with the screenshot, it's not clear to me what you decided to do with the search bar for the ByDay screen
> * I see that in the UI/UX fixes you made the search bar from the Search screen a separate composable, do you plan to reuse it for trips and activities? To me it would seem nicer if the search bar looks like the one on the Search screen for consistency, but depends what you and @1andreea26 think as well
> * please solve the merge conflicts
> * the last commit didn't go through the CI, this happens sometimes, but it will be fixed once you push the merge changes. It's important that we have a look at the test coverage as well.
> 
> Overall, the search bar works well, thanks for implementing this feature and taking into consideration our suggestions! I'll let Andreea give you the approval once she has a look too.

Thank you for the detailed feedback! 

I apologize for missing your comment with the screenshot. Regarding the ByDay screen, I thought it might be redundant to add both the search bar and filter there, but I'm open to implementing it if you think it would be valuable. Perhaps we could address this in a separate PR to keep the current changes focused?

And yes, you caught my intention with the SearchBar component! I made it a separate composable with plans to reuse it across trips and activities screens. I kept the styling minimal for now since I was planning to align all search bars to match the Search screen's design for consistency. I'd love to hear yours and @1andreea26's thoughts on this approach.

Thanks for the heads up about the CI issue! I'll get those merge conflicts resolved and make sure everything passes through CI properly this time :)

**Comment by ioanajianu** [comment]
> > Thanks for implementing our suggestions! I have a few comments to make:
> > 
> > * nice work in reusing the `NoResultsFound` composable!
> > * the padding looks better now, thanks for fixing that
> > * you didn't reply to my comment with the screenshot, it's not clear to me what you decided to do with the search bar for the ByDay screen
> > * I see that in the UI/UX fixes you made the search bar from the Search screen a separate composable, do you plan to reuse it for trips and activities? To me it would seem nicer if the search bar looks like the one on the Search screen for consistency, but depends what you and @1andreea26 think as well
> > * please solve the merge conflicts
> > * the last commit didn't go through the CI, this happens sometimes, but it will be fixed once you push the merge changes. It's important that we have a look at the test coverage as well.
> > 
> > Overall, the search bar works well, thanks for implementing this feature and taking into consideration our suggestions! I'll let Andreea give you the approval once she has a look too.
> 
> Thank you for the detailed feedback!
> 
> I apologize for missing your comment with the screenshot. Regarding the ByDay screen, I thought it might be redundant to add both the search bar and filter there, but I'm open to implementing it if you think it would be valuable. Perhaps we could address this in a separate PR to keep the current changes focused?
> 
> And yes, you caught my intention with the SearchBar component! I made it a separate composable with plans to reuse it across trips and activities screens. I kept the styling minimal for now since I was planning to align all search bars to match the Search screen's design for consistency. I'd love to hear yours and @1andreea26's thoughts on this approach.
> 
> Thanks for the heads up about the CI issue! I'll get those merge conflicts resolved and make sure everything passes through CI properly this time :)

Thanks for the reply!

By the ByDay screen redundancy, do you mean that during one day there won't be enough activities to search or filter by? I agree that might be true, but I was thinking that it could be consistent with the Activities screen, and I don't think it hurts to have more filtering functionality. Wdyt? Also @1andreea26, lmk your input on this

Glad to hear about the search bar composable, thanks for implementing that!

**Comment by 1andreea26** [comment]
> > Thanks for implementing our suggestions! I have a few comments to make:
> > 
> > * nice work in reusing the `NoResultsFound` composable!
> > * the padding looks better now, thanks for fixing that
> > * you didn't reply to my comment with the screenshot, it's not clear to me what you decided to do with the search bar for the ByDay screen
> > * I see that in the UI/UX fixes you made the search bar from the Search screen a separate composable, do you plan to reuse it for trips and activities? To me it would seem nicer if the search bar looks like the one on the Search screen for consistency, but depends what you and @1andreea26 think as well
> > * please solve the merge conflicts
> > * the last commit didn't go through the CI, this happens sometimes, but it will be fixed once you push the merge changes. It's important that we have a look at the test coverage as well.
> > 
> > Overall, the search bar works well, thanks for implementing this feature and taking into consideration our suggestions! I'll let Andreea give you the approval once she has a look too.
> 
> Thank you for the detailed feedback!
> 
> I apologize for missing your comment with the screenshot. Regarding the ByDay screen, I thought it might be redundant to add both the search bar and filter there, but I'm open to implementing it if you think it would be valuable. Perhaps we could address this in a separate PR to keep the current changes focused?
> 
> And yes, you caught my intention with the SearchBar component! I made it a separate composable with plans to reuse it across trips and activities screens. I kept the styling minimal for now since I was planning to align all search bars to match the Search screen's design for consistency. I'd love to hear yours and @1andreea26's thoughts on this approach.
> 
> Thanks for the heads up about the CI issue! I'll get those merge conflicts resolved and make sure everything passes through CI properly this time :)

Thank you Eleonora for implementing the changes we suggested! Looks good! 
- Regarding the ByDay screen search bar and filter, I think the search bar would be a nice addition, although I did think about adding a filter back when I was implementing the filter button, but I was wondering how to filter, to search for a day or for an activity? And if we search for activities, should we display all days in which the search activities are found? I think I need your input on this @ioanajianu , but I totally think it would help with consistency and add a nice feature! 
- Moreover, I think the previous search bar implementation on the Overview Screen was better, with the small icon in the right side, but I see your point in adding a search bar component to use in all of the screens (could reduce significantly our duplicated code). I agree with leaving it like this.

**Comment by ioanajianu** [comment]
> * Regarding the ByDay screen search bar and filter, I think the search bar would be a nice addition, although I did think about adding a filter back when I was implementing the filter button, but I was wondering how to filter, to search for a day or for an activity? And if we search for activities, should we display all days in which the search activities are found? I think I need your input on this @ioanajianu , but I totally think it would help with consistency and add a nice feature!

So the ByDay screen already filters the activities by a certain date (the one on which you press on in the schedule view - There's a ByDay screen for every possible day in the trip). So I think the filter would filter activities only by type.

**Comment by 1andreea26** [comment]
> > * Regarding the ByDay screen search bar and filter, I think the search bar would be a nice addition, although I did think about adding a filter back when I was implementing the filter button, but I was wondering how to filter, to search for a day or for an activity? And if we search for activities, should we display all days in which the search activities are found? I think I need your input on this @ioanajianu , but I totally think it would help with consistency and add a nice feature!
> 
> So the ByDay screen already filters the activities by a certain date (the one on which you press on in the schedule view - There's a ByDay screen for every possible day in the trip). So I think the filter would filter activities only by type.

But so, it would display the days with activities of the filtered type, right?

**Comment by ioanajianu** [comment]
> But so, it would display the days with activities of the filtered type, right?

The ByDay screen doesn't display days, only activities with a certain day. So, first you select a certain day, then you select what filter to apply (by type, since the day is set by definition of the ByDay screen)

**Comment by 1andreea26** [comment]
> > But so, it would display the days with activities of the filtered type, right?
> 
> The ByDay screen doesn't display days, only activities with a certain day. So, first you select a certain day, then you select what filter to apply (by type, since the day is set by definition of the ByDay screen)

I see what you mean now, but I think you're referring to the `ActivitiesForOneDay `screen (which the user accesses when they press on a card in the `ByDay `screen , that's why I was confused. In this case, yes I think this would be a nice addition, and since @eleonoraka  implemented a common search composable, I think she can add it there as well, in her next PR. In this case, after fixing the CI, I will approve :)

**Review by ioanajianu — APPROVED** [review]
Thanks Eleonora for implementing this feature! I tested it and it works well.

Apologies for the confusion with the ByDay screen, I did indeed mean ActivitiesForOneDay screen as Andreea mentioned. As we discussed, feel free to add the search bar there in another PR :)

Also, if you would like, let us know when you implement the common search bar across all the screens, I would be interested to know how it looks like.

The test coverage is above 80%, great work! Sorry if this review took longer than usual - now that the app is progressing, I feel the UI is becoming increasingly important to get just right.

LGTM :)

**Comment by eleonoraka** [comment]
> Thanks Eleonora for implementing this feature! I tested it and it works well.
> 
> Apologies for the confusion with the ByDay screen, I did indeed mean ActivitiesForOneDay screen as Andreea mentioned. As we discussed, feel free to add the search bar there in another PR :)
> 
> Also, if you would like, let us know when you implement the common search bar across all the screens, I would be interested to know how it looks like.
> 
> The test coverage is above 80%, great work! Sorry if this review took longer than usual - now that the app is progressing, I feel the UI is becoming increasingly important to get just right.
> 
> LGTM :)

Thanks for the thorough review! The search bar is in the UI/UX fixes PR right now. 
No worries about the review time - getting the UI right is important as we grow! Let me know if you need anything else.

**Review by 1andreea26 — APPROVED** [review]
I will approve as well, everything looks good and we've clarified everything :) Nice work, Eleonora! Don't forget to solve the merge conflicts and delete the branch after merging!

**Pull request merged**
